segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+                            failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> readEvents() {
+            AtomicLong startOffset = new AtomicLong(0L);
+            AtomicInteger readLength = new AtomicInteger(0);
+            return CompletableFuture.supplyAsync(() -> {
+                                        // Need to get the start and end offsets to properly create async read request.
+                                        startOffset.set(segment.getInfo().getStartOffset());
+                                        readLength.set((int) Math.min(segment.getInfo().getLength() - startOffset.get(), MAX_TOTAL_EVENT_SIZE));
+                                        return segment.read(startOffset.get(), readLength.get(), containerOperationTimeout);
+                                    }, executor)
+                                    .thenComposeAsync(rr -> AsyncReadResultProcessor.processAll(rr, executor, containerOperationTimeout), executor)
+                                    .thenApply(inputData -> deserializeEvents(inputData, startOffset.get(), readLength.get()));
+        }
+
+        private EventsReadAndTruncationPoints deserializeEvents(BufferView inputData, long startOffset, int maxLength) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long nextOffset = startOffset;
+            final long maxOffset = startOffset + maxLength;
+            BufferView.Reader input = inputData.getBufferViewReader();
+            try {
+                while (nextOffset < maxOffset && events.size() < getConfig().getMaxItemsAtOnce()) {
+                    ProcessorEventData event = ProcessorEventSerializer.deserializeEvent(input);
+                    events.add(event);
+                    // Update the offset to the beginning of the next event.
+                    nextOffset += HEADER_LENGTH + event.getLength();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            }
+
+            if (events.isEmpty()) {
+                throw new NoDataAvailableException();
+            }
+            return new EventsReadAndTruncationPoints(events);
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> applyProcessorHandler(EventsReadAndTruncationPoints readResult) {
+            return getHandler().apply(readResult.getProcessorEventsData()).thenApply(v -> readResult);
+        }
+
+        private CompletableFuture<Void> truncateInternalSegment(EventsReadAndTruncationPoints readResult, Timer iterationTime) {
+            long truncationOffset = this.segment.getInfo().getStartOffset() + readResult.getReadBytes();
```

Comment:
readResult.getReadBytes();

again same question - what happens with partial events. (event data that is spread across multiple reads)

Reply:
The only events in a `EventsReadAndTruncationLength` object are valid ones (not partial, as a partial read is expected to trigger an exception).